src/api/libcellml/generator.h

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright libCellML Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+
+#include <algorithm>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "libcellml/component.h"
+#include "libcellml/logger.h"
+#include "libcellml/model.h"
+#include "libcellml/operatorlibrary.h"
+#include "libcellml/variable.h"
+#include "../xmldoc.h"
+#include "../xmlnode.h"
+
+//! Everything in libCellML is in this namespace.
+namespace libcellml {
+
+#ifdef _MSC_VER
+    #pragma warning(push)
+    #pragma warning(disable: 4275)
+#endif
+struct LIBCELLML_EXPORT CodeNotGenerated : public std::exception
+{
+	const char * what () const throw ();
+};
+#ifdef _MSC_VER
+    #pragma warning(pop)
+#endif
+
+#ifdef _MSC_VER
+    #pragma warning(push)
+    #pragma warning(disable: 4275)
+#endif
+struct LIBCELLML_EXPORT UnknownNode : public std::exception
+{
+    const char * what () const throw ();
+};
+#ifdef _MSC_VER
+    #pragma warning(pop)
+#endif
+
+struct EnumClassHash
+{
+    template <typename T>
+    std::size_t operator()(T t) const
+    {
+        return static_cast<std::size_t>(t);
+    }
+};
+
+class LIBCELLML_EXPORT CXX
+{
+public:
+    enum class types {void_t, double_t, double_ct, double_pt, double_rt};
+
+    static std::string returnType(types t);
+    static std::string argType(types t);
+    static std::string argListOp() {return "(";}
+    static std::string argListCl() {return ")";}
+    static std::string funBodyOp() {return "{";}
+    static std::string funBodyCl() {return "}";}
+    static std::string instructionDelimiter() {return ";";}
+    static std::string dereferenceOp() {return "*";}
+};
+
+class LIBCELLML_EXPORT Generator : public Logger
+{
+public:
+    Generator();
+    virtual ~Generator();
+
+    std::string generateCode(ModelPtr m);
+    void writeCodeToFile(std::string filename);
+
+private:
```

Comment:
When using the Pimpl idiom we shouldn't have any private methods visible in a public API header file.  These private methods should all be moved into the pimpl class in the .cpp file.

Reply:
I think I solved this.